apple/swift-docc
# Correct source code locations of link diagnostics for Objective-C doc comments

**Explanation**: Fix the source location for some Obj-C diagnostics.
**Scope**: Fixes the line and column numbers that appear in the diagnostics.json file, used by Xcode to determine where to show them in the editor.
**GitHub Issue**: none.
**Risk**: Low. Minor code changes impacting diagnostic source locations only..
**Testing**: Ran the ./bin/test script and it succeeded
**Reviewer**: @d-ronnqvist


Bug/issue #, if applicable: 

rdar://102855081

## Summary

Correct the line and character locations of invalid link diagnostics in Objective-C source code doc comments.

Previously, `DocumentationContext#resolveLinks` would use the default symbol to calculate the offset of each doc comment in a source file. However, the default symbol is always set to a Swift symbol even for Objective-C source code. This lead to a bug when Objective-C diagnostics would always use a zero offset.

Example:

![Example of an Objective-C link diagnostic at the wrong location](https://github.com/apple/swift-docc/assets/28140/a0909448-5f6a-4366-92ce-0cd32436ac8b)

Instead, save the offset of each doc comment in the `DocumentationChunk` structure, which is initialized using the documented symbol, i.e. the Swift or Objective-C symbol which actually contains the doc comment.

Also decrement the line and character indexes by 1 for both Objective-C doc comments so Xcode displays them in the proper location.

Same example with the fix in place:

![Example of an Objective-C link diagnostic at the wrong location](https://github.com/apple/swift-docc/assets/28140/fb45558b-593b-4b88-9e15-c5dfbd315dc0)

## Dependencies

None.

## Testing

Introduced invalid links into doc comments inside of various Objective-C header files, and built the documentation. Tested that Xcode displayed the diagnostics in the proper location in the source code editor.

Steps:
1. Add ``invalid-link`` or some other invalid DocC link in a doc comment in an Objective-C header file.
2. Build the documentation using Xcode.
3. Check that the yellow warning diagnostic appears in the proper location. See the screen shots above.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [ x ] Added tests
- [ x ] Ran the `./bin/test` script and it succeeded
